File: 0257-binary-tree-paths/0257-binary-tree-paths.py
# Definition for a binary tree node.
# class TreeNode:
#   def __init__(self, val=0, left=None, right=None):
#     self.val = val
#     self.left = left
#     self.right = right
class Solution:
  def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
    if not root.left and not root.right:
      return [str(root.val)]
    ans = []
    if root.left:
      ans += [f"{str(root.val)}->{path}" for path in self.binaryTreePaths(root.left)]
    if root.right:
      ans += [f"{str(root.val)}->{path}" for path in self.binaryTreePaths(root.right)]
    return ans

File: 0257-binary-tree-paths/0257-binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *   int val;
 *   TreeNode left;
 *   TreeNode right;
 *   TreeNode() {}
 *   TreeNode(int val) { this.val = val; }
 *   TreeNode(int val, TreeNode left, TreeNode right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 *   }
 * }
 */
class Solution {
  public List<String> binaryTreePaths(TreeNode root) {
    List<String> ans = new ArrayList<String>();
    if (root.left == null && root.right == null) {
      ans.add(String.valueOf(root.val));
    } else {
      if (root.left != null) {
        List<String> leftPaths = binaryTreePaths(root.left).stream()
          .map(path -> String.valueOf(root.val) + "->" + path)
          .collect(Collectors.toList());
        ans.addAll(leftPaths);
      }
      if (root.right != null) {
        List<String> rightPaths = binaryTreePaths(root.right).stream()
          .map(path -> String.valueOf(root.val) + "->" + path)
          .collect(Collectors.toList());
        ans.addAll(rightPaths);
      }
    }
    return ans;
  }
}